fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,509 @@ getinfo_test(caps, 4, "Test for capability bit regression",
+	 * rx_attr::caps are zero, these returns caps should be set based on
+	 * fi_info::caps
+	 */
+	hints->caps = FI_TAGGED;
+	hints->optional_caps = FI_ATOMIC;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf,
+			       "fi_getinfo with optional_caps failed to return",
+			       ret);
+		return -FI_EINVAL;
+	}
+
+	ret = test_valid_optional_cap_info(*info, FI_ATOMIC, FI_ATOMIC,
+					   FI_ATOMIC);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "Optional caps incorrectly returned",
+			       ret);
+		return ret;
+	}
+
+	/* For RxM, FI_ATOMIC used as optional_cap fi_getinfo() input should not
+	 * be returned in caps output since performance could be impacted.
+	 */
+	ret = test_invalid_cap_info(*info, FI_ATOMIC, FI_ATOMIC, FI_ATOMIC);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf,
+			       "FI_ATOMIC incorrectly returned in caps",
+			       ret);
+		return ret;
+	}
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Verify that using domain_attr::optional_caps, tx_attr::optional_caps,
+	 * and rx_attr::optional_caps works as expected.
+	 */
+	hints->caps = required_caps;
+	hints->optional_caps = 0;
+	hints->tx_attr->caps = FI_TAGGED;
+	hints->rx_attr->caps = FI_TAGGED;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf,
+			       "fi_getinfo with unexpectedly failed to return",
+			       ret);
+		return -FI_EINVAL;
+	}
+
+	ret = test_valid_optional_cap_info(*info, 0, 0, 0);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "Optional caps incorrectly returned",
+			       ret);
+		return ret;
+	}
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	hints->tx_attr->optional_caps = FI_ATOMIC;
+	hints->rx_attr->optional_caps = FI_ATOMIC;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf,
+			       "fi_getinfo with optional_caps failed to return",
+			       ret);
+		return -FI_EINVAL;
+	}
+
+	ret = test_valid_optional_cap_info(*info, 0, FI_ATOMIC, FI_ATOMIC);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "Optional caps incorrectly returned",
+			       ret);
+		return ret;
+	}
+
+	ret = test_invalid_cap_info(*info, 0, FI_ATOMIC, FI_ATOMIC);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf,
+			       "FI_ATOMIC incorrectly returned in caps",
+			       ret);
+		return ret;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int test_ignore_optional_cap(char *node, char *service, uint64_t flags,
+				    struct fi_info *hints,
+				    struct fi_info **info)
+{
+	uint64_t required_caps = FI_MSG | FI_TAGGED;
+	int ret;
+
+	/* Verify that the provider meets the required capabilities of this
+	 * test.
+	 */
+	hints->caps = required_caps;
+	ret = fi_getinfo(FI_VERSION(1, 19), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		sprintf(err_buf, "No providers found");
+		return -FI_ENOSYS;
+	}
+
+	hints->domain_attr->name = strdup((*info)->domain_attr->name);
```

Comment:
This strdup is causing windows debug job to fail. Allocation is coming from one heap and being freed on another. Since it is not strdup is not strictly needed for this test, I'll remove it.